Add tutorial for implementing shopping cart (#32)
Genez-io/genez-io.github.io: content/posts/implement_shopping_cart.md

Let's get started! 🚀

# Content
- [Introduction](#introduction)
- [Prerequisites](#prerequisites)
- [Getting Started](#getting-started)
- [Setting up the backend](#setting-up-the-backend)
- [Setting up the frontend](#setting-up-the-frontend)
- [Deploying the application](#deploying-the-application)
- [Improvements](#improvements)
  - [Add a counter for the number of items in the cart](#add-a-counter-for-the-number-of-items-in-the-cart)
  - [Add spinners animation for loading states](#add-spinners-animation-for-loading-states)
- [Conclusion](#conclusion)

## Introduction

A possible implementation of a shopping cart is to store the items as a cookie on the client-side. The disadvantage of this approach is that the cookie is attached to every request and large cookies can significantly slow down your application.

A better option is to use a Redis database on the server-side to store items persistently and retrieve them very fast when needed.

This tutorial will show you how to harness the speed of a Redis database to build a shopping cart functionality using TypeScript/NodeJs.

{{< details "Expand this section to get more information on what's Redis" >}}

> Redis is an open-source, in-memory database that store key, values pairs. Redis is a highly versatile and performant database system. One of Redis's key features is its speed, as it primarily operates in-memory, enabling rapid data access. It is widely used for caching, real-time analytics, message brokering, and other applications where low-latency and high-throughput data storage and retrieval are crucial. Redis also provides persistence options, allowing data to be saved to disk for durability. Its simplicity, efficiency, and support for advanced data structures make Redis a popular choice for applications requiring fast and scalable data storage and retrieval.

Reviewer: ```suggestion
Redis is an open-source, in-memory database that stores key, values pairs. Redis is a highly versatile and performant database system. One of Redis's key features is its speed, as it primarily operates in-memory, enabling rapid data access. It is widely used for caching, real-time analytics, message brokering, and other applications where low-latency and high-throughput data storage and retrieval are crucial. Redis also provides persistence options, allowing data to be saved to disk for durability. Its simplicity, efficiency, and support for advanced data structures make Redis a popular choice for applications requiring fast and scalable data storage and retrieval.
```

The end result of this tutorial will be a web application that looks like this:

![Alt text](/images/blog/shopping_cart_example/shopping_cart_preview.png)

## Prerequisites

If you don't already have them, you'll need to install the following tools:
- [Node.js](https://nodejs.org/en/download/current) (version >= 16.0.0)
- [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
- [Genezio](https://genez.io)

Note: I recommend you to use [nvm](https://github.com/nvm-sh/nvm#installing-and-updating) to manage NodeJs and npm versions.
After installing `nvm`, you can easily get the any version of `node` by running `nvm install <node_version>`.
`nvm` will automatically install the corresponding `npm` version.

## Getting Started

Firstly, you'll need to create a new project with a server and a client.
The server will be responsible to communicate (store, retrieve, update, delete) with the Redis database.
The client will render the products available in the store and allow the user to add them to the shopping cart.

To get started with a template, install `genezio` using `npm` and run it in your terminal.

> Later on, `genezio` come in handy to deploy and host your web applications in the cloud.

Reviewer: Do you mean `genezio will come in handy`?

```bash
npm install -g genezio
```

After installing `genezio`, you can scaffold the application by running the following command:

```bash
genezio
```

The command above will get you through a series of questions to help you customize and prepare your project for production deployment.

> Your terminal should similar to the following output:

Reviewer: ```suggestion
Your terminal should look similar to the following output:
```

```
~ genezio
Redirecting to browser to complete authentication...

? Choose a template for your genezio project Fullstack
Your project will start from the Fullstack template.

? Please enter a name for your project: my-online-store
Your project will be named my-online-store.

? Choose a region for your project US East (N. Virginia)
Your project will be deployed in US East (N. Virginia).

We are creating the project in the current directory.

Deploying your backend project to genezio infrastructure...

Your backend project has been deployed and is available at https://app.genez.io/project/c7e163b7-d0c9-47f6-8520-0e546f2e7b07/cdfc6daf-2c9d-4fae-9c94-4f011c8a84f2

Deploying your frontend to genezio infrastructure...

No subdomain specified in the genezio.yaml configuration file. We will provide a random one for you.
Frontend successfully deployed at https://amber-cooperative-blackbird.app.genez.io.
```

The output from the `genezio` command will contain a randomly-assigned subdomain where your project was deployed for testing purposes.

## Setting up the backend

> Now that you have a template to start from, fire up your favorite code editor and open project's directory.

Reviewer: ```suggestion
Now that you have a template to start from, fire up your favorite code editor and open the project's directory.
```

    // check if cart is empty
    if (Object.keys(cart).length === 0) {
      console.error("Cart is empty");
      return [];
    }

    // map cart items to an array of CartItem objects
    const cartItems: CartItem[] = [];
    for (const key in cart) {
      const item = {
        title: key,
        count: cart[key],
      };
      cartItems.push(item);
    }

    return cartItems;
  }
}
```

Install the dependencies by running the following command in the `server` directory:

```bash
npm install ioredis
npm --save-dev install @types/node
```

Right now, the backend won't work properly because there is no `UPSTASH_REDIS_URL` environment variable set in the `.env`.
Let's create a Redis database and connect it to your web application using a `.env` file.

Luckily, genezio makes it very easy to add an Upstash Redis integration to your project.
Go to the [genezio dashboard](https://app.genez.io) and add an Upstash Redis integration to your project.

![Alt text](/images/blog/shopping_cart_example/add_integration.png)

If you get stuck at this step, you can follow the [Upstash Redis integration tutorial](https://docs.genez.io/genezio-documentation/integrations/upstash-redis).

> After adding the Upstash Redis integration, copy the environment variable from the `.env` section. Go back to your `server` directory, create a new file called `.env` and paste the contents there.

Reviewer: I think somebody might think that they should copy the *** star output and paste it into the `.env` file. You should mention that they should explicitly use the copy button to disclose their tokens.

![Alt text](/images/blog/shopping_cart_example/copy_env_vars.png)

To test the code, you can run `genezio local` in the project directory.
This will start a fast local development server on your machine.

```bash
genezio local
```

You can test your application from the genezio dashboard by following the URL exposed in the terminal by `genezio local`.

```
~ genezio local
Test your code at https://app.genez.io/test-interface/local?port=8083
```

> From the genezio dashboard you can sent request to your backend classes and see the response.

Reviewer: ```suggestion
From the genezio dashboard you can send requests to your backend classes and see the responses.
```

![Alt text](/images/blog/shopping_cart_example/test_backend.png)

> You'll also need 2 more methods in the backend source code - a method to remove only an item from the cart and a method to delete the entire cart.

Reviewer: ```suggestion
 You'll also need 2 more methods in the backend source code - a method to remove only one item from the cart and a method to delete the entire cart.
```

          setCartData(cart);
        } catch (error) {
          console.error('Error fetching or parsing cart data:', error);
        }
      };

      fetchCartData();
    }
  }, [isCartVisible, token]);
```

Send a request to the backend to delete an item from the cart when the `Delete` button is clicked.

```typescript
  const handleDeleteItem = async (e: any, cartItem: CartItem) => {
    e.preventDefault();

    await ShoppingCartService.removeItemFromCart(token, cartItem.title);

    // You can implement your delete logic here
    const updatedCartData = await ShoppingCartService.getCart(token);
    setCartData(updatedCartData);
  };
```

Send a request to the backend to clear the cart when the `Clear Cart` button is clicked.

```typescript
  const handleClearCart = async (e: any) => {
    e.preventDefault();

    // You can implement your clear cart logic here
    await ShoppingCartService.deleteCart(token);

    // Update the cart data
    const updatedCartData = await ShoppingCartService.getCart(token);
    setCartData(updatedCartData);
  };
```

## Deploying the application

Your application is now ready to be deployed to the cloud to be used by your clients.

> Before deploying your application, you can change the randomly-assigned subdomain to something more meaningly for your application.

Reviewer: ```suggestion
Before deploying your application, you can change the randomly-assigned subdomain to something more meaningful for your application.
```

  const handleBuyClick = async (e: any, product: Product) => {
    // ...

    // Update the local state to reflect the purchased quantity
    setPurchasedQuantity((prevQuantity) => prevQuantity + 1);

    // ...
  };

  const handleDeleteItem = async (e: any, cartItem: CartItem) => {
    // ...

    // Update the local state to reflect the purchased quantity
    setPurchasedQuantity((prevQuantity) => prevQuantity - 1);

    // ...
  };

  const handleClearCart = async (e: any) => {
    // ...

    // Update the local state to reflect the purchased quantity
    setPurchasedQuantity(0);

    // ...
  };

  return (
    <Button color="primary" onClick={(e) => toggleCartModal(e)}>
      <FaShoppingCart />
      {purchasedQuantity > 0 && (
        <span className="m-2">{purchasedQuantity}</span>
      )}
    </Button>
  );
```

This state should be updated each time an item is added or removed from the cart. When the cart is cleared, the state should be reset to 0.

### Add spinners animation for loading states

To improve the UI even further, you can add spinners to indicate loading states on the buttons that are sending requests to the backend.

> Each button should have a its own loading state to indicate that the request is being processed.

Reviewer: A determiner cannot be combined with a possessive pronoun. Did you mean simply 'a' or 'its'?

Below is an example of how to add a loading state to the `Buy now` button.

```typescript
  const [addItemLoading, setAddItemLoading] = useState<{
    [key: number]: boolean;
  }>({});

  const handleBuyClick = async (e: any, product: Product) => {
    // Set loading state to true to show the spinner
    setAddItemLoading((prevStates) => ({ ...prevStates, [product.id]: true }));

    // ...

    // Set loading state to false to hide the spinner
    setAddItemLoading((prevStates) => ({ ...prevStates, [product.id]: false }));
  };

  return (
    <Button
      color="primary"
      onClick={(e) => handleBuyClick(e, product)}
      disabled={addItemLoading[product.id]}
    >
      {addItemLoading[product.id] ? (
        <Spinner size="sm" color="light" />
      ) : (
        'Buy Now'
      )}
    </Button>
);
```

## Conclusion

Congratulations! 🥳

> The codebase for this tutorial is open-source and you can find it on the [GitHub repository](https://github.com/genez-io/genezio-examples/tree/main/typescript/shopping-cart).

Reviewer: ```suggestion
The codebase for this tutorial is open-source, and you can find it on the [GitHub repository](https://github.com/genez-io/genezio-examples/tree/main/typescript/shopping-cart).
```
